enumerables.rb

```diff
@@ -0,0 +1,175 @@
+# module custom Enumerables
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  # My Each Method
+
+  def my_each
+    return to_enum unless block_given?
+
+    length.times do |i|
+      yield(self[i])
+    end
+  end
+
+  # My  Each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    length.times do |i|
+      yield(i, self[i])
+    end
+  end
+
+  # My Select Method
+  def my_select
+    return to_enum unless block_given?
+
+    invited_list = []
+    my_each do |friend|
+      invited_list.push(friend) if yield(friend)
+    end
+
+    invited_list
+  end
+
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  # My  All
+  def my_all?(args = nil)
+    if block_given?
+      my_each do |element|
+        return false if yield(element) == false
+      end
+    elsif args.nil?
+      my_each do |i|
+        return false if i.nil?
+      end
+    elsif args.nil? && (args.is_a? Class)
+      return false
+    elsif !args.nil? && (args.is_a? Class)
+      return true
+    elsif Regexp
+      my_each do |i|
+        return false if args.match(i)
+      end
+    end
+    true
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def my_any?(args = nil)
+    check = false
+    is_true = false
+    if block_given?
+      my_each do |i|
+        if yield(i) == true
+          check = true
+          return check
+        else
+          check = false
+        end
+      end
+      return check
+    elsif args.class == Regexp
+      my_each do |i|
+        return true if i.match(args)
+
+        return false
+      end
+    elsif !args.nil? && (args.is_a? Class)
+      my_each do |i|
+        if i.is_a? args
+          is_true = true
+          return true
+        else
+          is_true = false
+        end
+      end
+      return is_true
+    end
+    unless block_given?
+      my_each { |x| return true if x == true }
+      return false
+    end
+    false
+  end
+
+  def my_none?(args = nil)
+    if block_given?
+      check = false
+      my_each do |element|
+        if yield(element) == false
+          check = true
+        else
+          check = false
+          return check
+        end
+      end
+      return check
+    elsif !args.nil? && (args.is_a? Class)
+      my_each do |i|
+        return false if i.is_a? args
+      end
+    elsif args.class == Regexp
+      my_each do |i|
+        return false if i.match(args)
+
+        return true
+      end
+    end
+    unless block_given?
+      my_each { |x| return false if x == true }
+      return true
+    end
+    false
+  end
+
+  # rubocop:enable Metrics/MethodLength
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  # My Count
+  def my_count(num = 0)
+    return size unless block_given? || num.zero?
+
+    counter = 0
+    my_each do |i|
+      counter += 1 if yield(i)
+    end
+    counter
+  end
+
+  # My Inject
+  def my_inject
```

Comment:
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
```
- [ ] raises a "LocalJumpError" when no block or argument is given
```ruby
array.my_inject == raise_error(LocalJumpError) # should return true
```
- [ ] searches for the longest word in an array of strings
```ruby
words = %w[dog door rod blade] 
words.my_inject(&search) == words.inject(&search) # should return true
```
- [ ] does not mutate the original array
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject { |num| num + 1 }
array == array_clone # should return true
```
- [ ] when a block is given without an initail value
```ruby
array.my_inject(&operation) == array.inject(&operation)
```